### DISCUSSION TITLE
Homebrew asks me to generate a PAT

### DISCUSSION THREAD

**Discussion by jketema** [Homebrew/discussions#3232] [Everyday usage]
Hi,

 Running `brew search` gives me the following warning:
```
Warning: Error searching on GitHub: GitHub API Error: Bad credentials
The GitHub credentials in the macOS keychain may be invalid.
Clear them with:
  printf "protocol=https\nhost=github.com\n" | git credential-osxkeychain erase
Create a GitHub personal access token:
    https://github.com/settings/tokens/new?scopes=gist,repo,workflow&description=Homebrew
  echo 'export HOMEBREW_GITHUB_API_TOKEN=your_token_here' >> ~/.zshrc
```
Two questions:
* Why am I asked to generate a PAT? I thought all Homebrew repositories were public? This might just need better explanation in the warning message.
* Following the link let me generate a PAT that gives access to all my private repositories without the option of turning this off. This is a big nope, as brew is untrusted in this respect. Why is this necessary?

**Comment by SMillerDev** [comment]
It suggests this because the PAT you have set is invalid. You can also just remove it and homebrew will still work fine.

However, since you probably set it for a reason it also suggests a way to create a new PAT

**Reply by jketema** [reply]
I never set a PAT. I did find a stale password, so the warning is very deceptive.

**Reply by jketema** [reply]
It's also weird that Homebrew simply assumes that the stale password/PAT had something to do with how I interact with Homebrew. It was definitely there for some other reason.

**Reply by SMillerDev** [reply]
Homebrew is essentially all git(hub) underneath so it'll use whatever information you use to authenticate with GitHub.

It's very hard for homebrew to determine why credentials are set, it just knows it does or does not have access when it tries to connect

**Reply by jketema** [reply]
Except that all of Homebrew is public, and hardly anyone (if anyone at all) will directly modify the repos in the Homebrew org, so assuming any kind of credentials are needed is just weird/misleading. The PAT suggestion in the warning will give Homebrew access to things it probably shouldn't have access to, so this is just an extremely bad way being "helpful" for the unsuspecting user.

**Reply by SMillerDev** [reply]
There are a lot of companies using private repos as homebrew sources and there is a reasonable chance of being rate limited if you don't use some sort of authentication with GitHub. The warning is there because people complained about this when homebrew didn't suggest it.

I'm not sure what scope is needed to interact with homebrew for most people, but if you think you do a pull request would be most welcome.

**Reply by jketema** [reply]
> There are a lot of companies using private repos as homebrew sources and there is a reasonable chance of being rate limited if you don't use some sort of authentication with GitHub.

Thanks that at least explains part of what is going on here, but then I would expect:
```
https://github.com/settings/tokens/new?scopes=repo&description=Homebrew
```
but it says:
```
https://github.com/settings/tokens/new?scopes=gist,repo,workflow&description=Homebrew
```
Why is it necessary to be able to create gists and update Action workflows?

**Reply by carlocab** [reply]
For most usage, it is not necessary. However, some commands require gist-scope (e.g. `brew gist-logs`), and some other commands (primarily developer commands, e.g. fetching built bottles from a PR workflow run with `brew pr-pull`) require workflow-scope.

**Comment by noelforte** [comment]
Ran into this issue today where like @jketema, `brew` is suddenly asking me for a PAT when it never has before. I'm not sure whether this is an issue specific to my machine or something going on with `brew` internals but either way I can't find the source of the credential issue.

`brew` outputs the following (which is slightly different than the original post, error is `Requires authentication` as opposed to `Bad credentials`):

```console
Warning: Error searching on GitHub: GitHub API Error: Requires authentication
The GitHub credentials in the macOS keychain may be invalid.
Clear them with:
  printf "protocol=https\nhost=github.com\n" | git credential-osxkeychain erase
Create a GitHub personal access token:
https://github.com/settings/tokens/new?scopes=gist,repo,workflow&description=Homebrew
echo 'export HOMEBREW_GITHUB_API_TOKEN=your_token_here' >> ~/.zshrc
```

What I've tried:
1. Clearing all "github" related items from the keychain
2. Running `printf "protocol=https\nhost=github.com\n" | git credential-osxkeychain erase` as suggested by `brew`
3. Uninstalling/reinstalling brew
4. Signing out of github everywhere (including the `gh` CLI) to clear any cached auth tokens
5. Relocating my ssh configs for github, thinking that maybe they were getting in the way

Nevertheless, even though there's no possible way that I feel like there's a token that homebrew is trying to use, `brew` continues to nag me to create a PAT.

Is there a specific reason why brew would suddenly ask for a PAT when it was working without throwing any warnings prior? Any suggestions for what to try next would be appreciated!

**Reply by noelforte** [reply]
Following up to my post above, I was able to reproduce the same issue on Linux, and have opened [Ticket 15208](https://github.com/Homebrew/brew/issues/15208). Leaving here in case someone else comes across it while investigating this behavior.

**Reply by karimbenbourenane** [reply]
Likewise, I am being affected by this issue today as well on an M2 Mac. Unusual that we'd have a cluster of 3 responses to this old issue in the span of 3 hours so I'm going to guess this may be a Github rate limiting thing that was changed recently, possibly in error.

**Reply by wdfry** [reply]
Same for me too. Fresh install about two days ago on new M2.

**Reply by icpd** [reply]
Same too. on M1Pro Mac

**Reply by jpguimaraes** [reply]
Same here, in a new M2

**Reply by maimuta8** [reply]
Happening for me on Ubuntu 22.04

**Reply by Bo98** [reply]
This will be fixed in Homebrew 4.0.13.

**Reply by darzok0914** [reply]
what's the best way to go around this problem in the meantime?

**Reply by maimuta8** [reply]
I'm still able to search. The results just come up after the above message.

**Reply by Bo98** [reply]
I've released 4.0.13 now. Run `brew update`, or wait for auto-update to pick it up.

**Reply by YusufGandhi** [reply]
thank you for fixing this @Bo98 . waking up this morning with the new errors. glad that I searched before generating PAT as suggested.

**Reply by git-jonathankinney** [reply]
Well, I don't use keychain in any of this, but I did notice this:

```bash
git config -l
# returns this in the top line of the output
credential.helper=osxkeychain
# I'm like "umm, no, if keychain is problematic for some unknown reason, then I am going to remove it, don't have time to debug why"
git config --system --unset credential.helper
# boom, problem gone.  Obviously not so simple if you have some need or use of that integration, but for me, no loss.
```
I can add it back to the ~/.gitconfig if I need.  Don't know if it persists, not sure if it was always there, and I don't see it in any of the git configs, so there are many unanswered questions, but I have things to do.
